R/labels.R

```diff
@@ -87,28 +102,141 @@ uses_labels <- function(p, ...) {
+    names(args)[default_labels], ~ unlist(default_label(p, .))
+  )
+
   if (!all(is_scalar_string_or_null(args))) {
     stop(
       "All inputs to `...` must be character vectors of length 1 or `NULL`.",
       call. = FALSE
     )
   }
 
-  if (is.null(names(args)) || any(names(args) == "")) {
-    stop("All inputs to `...` must be named.", call. = FALSE)
+  if (is.null(names(args))) {
+    names(args) <- rep("",    length(args))
+    named       <- rep(FALSE, length(args))
+  } else {
+    named <- names(args) != ""
   }
 
-  labels <- get_labels(p, names(args))
+  result         <- logical(length(args))
+  result[!named] <- check_labels_set(p, args[!named])
+  result[named]  <- check_labels_match(p, args[named])
+  names(result)  <- coalesce_chr(names(args), args)
+  result
+}
+
+#' What is the default label for a plot aesthetic?
+#'
+#' @family functions for checking labels
+#' @export
+#'
+#' @examples
+#' require(ggplot2)
+#'
+#' p <- ggplot(data = mpg, mapping = aes(x = displ, y = hwy)) +
+#'   geom_point(mapping = aes(color = class, shape = drv)) +
+#'   geom_smooth() +
+#'   labs(title = "My plot", x = "Weight", y = "MPG", color = NULL)
+#'
+#' # Returns the label the ggplot would create by default for an aesthetic
+#' default_label(p, "x")
+#' default_label(p, c("x", "y"))
+#' default_label(p)
+#'
+#' # If an aesthetic does not exist, returns NULL
+#' default_label(p, "size")
+#'
+#' # If an aesthetic has no default, returns NULL
+#' default_label(p, "title")
+#'
+#' # The colo(u)r aesthetic can be matched with or without a u
+#' default_label(p, "color")
+#' default_label(p, "colour")
+#' @param p A [ggplot][ggplot2::ggplot] object
+#' @param aes If `aes` is a [character] vector, returns only the labels
+#'   corresponding to the included aesthetics.
+#'   Defaults to [`NULL`], which returns all labels.
+#'
+#' @return A named [list] in which each element is a [character] string
+#'   or [`NULL`]
+default_label <- function(p, aes = NULL) {
+  UseMethod("default_label")
+}
+
+#' @export
+default_label.default <- function(p, aes = NULL) {
+  if (!missing(p)) {
+    rlang::abort("`p` must be a `ggplot` object.")
+  }
+
+  structure(list(), class = c(".default_label", "ggcheck_placeholder"))
+}
+
+#' @export
+default_label.ggplot <- function(p, aes = NULL) {
+  if (is.null(aes)) {
+    aes <- names(p$labels)
+  }
+
+  if (!is.character(aes)) {
+    rlang::abort("`aes` must be a character vector or NULL.")
+  }
+
+  names(aes) <- aes
+
+  aes[aes == "color"] <- "colour"
+
+  make_labels <- utils::getFromNamespace("make_labels", "ggplot2")
+
+  purrr::map(
+    aes,
+    function(aes) {
+      if (!is.null(p$mapping[[aes]])) {
+        return(as.character(make_labels(p$mapping[aes])))
+      }
+
+      for (i in seq_along(p$layers)) {
+        if (!is.null(p$layers[[i]]$mapping[[aes]])) {
+          return(as.character(make_labels(p$layers[[i]]$mapping[aes])))
+        }
+      }
+
+      NULL
+    }
```

Comment:
If this function replicates behavior in ggplot2, it'd be nice to have a link to the source code to reference in case implementations diverge.

Reply:
This is more of a reverse engineering of how `ggplot` works rather than a reimplementation of any of `ggplot`'s code, but 70ea762 adds comments explaining the logic of what's going on

Reply:
The added comments are perfect, thanks! Future versions of ourselves will greatly appreciate them 😄